pcsx2/GS/GSCapture.cpp

```diff
@@ -242,204 +346,531 @@ bool GSCapture::BeginCapture(float fps, GSVector2i recommendedResolution, float
-	if (!GSConfig.VideoCaptureCodec.empty())
-	{
-		codec = wrap_avcodec_find_encoder_by_name(GSConfig.VideoCaptureCodec.c_str());
-		if (!codec)
-		{
-			Host::AddIconOSDMessage("GSCaptureCodecNotFound", ICON_FA_CAMERA,
-				fmt::format("Video codec {} not found, using default.", GSConfig.VideoCaptureCodec),
-				Host::OSD_ERROR_DURATION);
-		}
-	}
-	if (!codec)
-		codec = wrap_avcodec_find_encoder(output_format->video_codec);
-	if (!codec)
-	{
-		Host::AddIconOSDMessage("GSCaptureError", ICON_FA_CAMERA, "Failed to find encoder.", Host::OSD_ERROR_DURATION);
-		EndCapture();
+		InternalEndCapture(lock);
 		return false;
 	}
 
 	int res = wrap_avformat_alloc_output_context2(&s_format_context, output_format, nullptr, s_filename.c_str());
 	if (res < 0)
 	{
 		LogAVError(res, "avformat_alloc_output_context2() failed: ");
-		EndCapture();
+		InternalEndCapture(lock);
 		return false;
 	}
 
-	s_codec_context = wrap_avcodec_alloc_context3(codec);
-	if (!s_codec_context)
+	// find the codec id
+	if (capture_video)
 	{
-		Host::AddIconOSDMessage("GSCaptureError", ICON_FA_CAMERA, "Failed to allocate codec context.", Host::OSD_ERROR_DURATION);
-		EndCapture();
-		return false;
-	}
+		const AVCodec* vcodec = nullptr;
+		if (!GSConfig.VideoCaptureCodec.empty())
+		{
+			vcodec = wrap_avcodec_find_encoder_by_name(GSConfig.VideoCaptureCodec.c_str());
+			if (!vcodec)
+			{
+				Host::AddIconOSDMessage("GSCaptureCodecNotFound", ICON_FA_CAMERA,
+					fmt::format("Video codec {} not found, using default.", GSConfig.VideoCaptureCodec), Host::OSD_ERROR_DURATION);
+			}
+		}
+		if (!vcodec)
+			vcodec = wrap_avcodec_find_encoder(output_format->video_codec);
+		if (!vcodec)
+		{
+			Host::AddIconOSDMessage("GSCaptureError", ICON_FA_CAMERA, "Failed to find video encoder.", Host::OSD_ERROR_DURATION);
+			InternalEndCapture(lock);
+			return false;
+		}
 
-	s_codec_context->codec_type = AVMEDIA_TYPE_VIDEO;
-	s_codec_context->bit_rate = GSConfig.VideoCaptureBitrate * 1000;
-	s_codec_context->width = s_size.x;
-	s_codec_context->height = s_size.y;
-	wrap_av_reduce(&s_codec_context->time_base.num, &s_codec_context->time_base.den,
-		10000, static_cast<s64>(static_cast<double>(fps) * 10000.0), std::numeric_limits<s32>::max());
+		s_video_codec_context = wrap_avcodec_alloc_context3(vcodec);
+		if (!s_video_codec_context)
+		{
+			Host::AddIconOSDMessage("GSCaptureError", ICON_FA_CAMERA, "Failed to allocate video codec context.", Host::OSD_ERROR_DURATION);
+			InternalEndCapture(lock);
+			return false;
+		}
 
-	// Default to YUV 4:2:0 if the codec doesn't specify a pixel format.
-	if (!codec->pix_fmts)
-	{
-		s_codec_context->pix_fmt = AV_PIX_FMT_YUV420P;
+		s_video_codec_context->codec_type = AVMEDIA_TYPE_VIDEO;
+		s_video_codec_context->bit_rate = GSConfig.VideoCaptureBitrate * 1000;
+		s_video_codec_context->width = s_size.x;
+		s_video_codec_context->height = s_size.y;
+		s_video_codec_context->sample_aspect_ratio = wrap_av_d2q(aspect, 100000);
```

Comment:
if i understand correctly, DAR value is wrongly set to SAR in this line

Reply:
It does, if the aspect ratio of the TV is not the same as the aspect ratio of the framebuffer (which is fairly common).